library/core/src/hash/mod.rs

```diff
@@ -602,7 +602,7 @@ pub trait BuildHasher {
 /// [`HashSet`]: ../../std/collections/struct.HashSet.html
 /// [zero-sized]: https://doc.rust-lang.org/nomicon/exotic-sizes.html#zero-sized-types-zsts
 #[stable(since = "1.7.0", feature = "build_hasher")]
-pub struct BuildHasherDefault<H>(marker::PhantomData<H>);
+pub struct BuildHasherDefault<H>(marker::PhantomData<fn() -> H>);
```

Comment:
@steffahn  just for curiosity, how does `fn() -> H` lifts the requirement, marking `BuildHasherDefault` `Send/Sync` no matter `H` Send/Sync or not?

Reply:
`PhantomData<T>` has e.g. [a `T: Send` bound/condition for `PhantomData<T>: Send`](https://doc.rust-lang.org/std/marker/struct.PhantomData.html#synthetic-implementations), while `fn() -> T` is unconditionally `Send` (and similarly for all the other auto-traits).

Auto-traits work by considering the types of the fields, hence `BuildHasherDefault<H>: Send` if and only if all it's field types are `Send`; i.e.  `BuildHasherDefault<H>: Send` used to require `PhantomData<H>: Send`, which in turn requires `H: Send`, while after this PR, it will require  `PhantomData<fn() -> H>: Send`, which in turn requires `fn() -> H: Send`, which is true unconditionally:
````
fn function_requiring_send<T: Send>() {}

fn demonstration<H /* not necessarily `Send */>() {
    function_requiring_send::<fn() -> H>(); // works!
}
````